classes/services/PKPPublicationService.inc.php

```diff
@@ -624,22 +652,39 @@ public function publish($publication)
 
         // Log an event when publication is published. Adjust the message depending
         // on whether this is the first publication or a subsequent version
-        if (count($submission->getData('publications')) > 1) {
-            $msg = $newPublication->getData('status') === PKPSubmission::STATUS_SCHEDULED ? 'publication.event.versionScheduled' : 'publication.event.versionPublished';
-        } else {
-            $msg = $newPublication->getData('status') === PKPSubmission::STATUS_SCHEDULED ? 'publication.event.scheduled' : 'publication.event.published';
-        }
-        SubmissionLog::logEvent(Application::get()->getRequest(), $submission, SubmissionEventLogEntry::SUBMISSION_LOG_METADATA_PUBLISH, $msg);
+        $availableMessages = [
+            0 => [
+                0 => 'publication.event.published',
+                1 => 'publication.event.scheduled',
+            ],
+            1 => [
+                0 => 'publication.event.versionPublished',
+                1 => 'publication.event.versionScheduled',
+            ],
+        ];
 
-        HookRegistry::call('Publication::publish', [&$newPublication, $publication, $submission]);
+        $msg = $availableMessages[(int) (count($submission->getData('publications')) > 1)][(int) ($newPublication->getData('status') === PKPSubmission::STATUS_SCHEDULED)];
+
+        SubmissionLog::logEvent(
+            Application::get()->getRequest(),
+            $submission,
+            SubmissionEventLogEntry::SUBMISSION_LOG_METADATA_PUBLISH,
+            $msg
+        );
+
+        HookRegistry::call(
+            'Publication::publish',
+            [
+                &$newPublication,
+                $publication,
+                $submission
+            ]
+        );
         event(new PublishedEvent($newPublication, $publication, $submission));
 
         // Update the search index.
-        if ($newPublication->getData('status') === PKPSubmission::STATUS_PUBLISHED) {
-            Application::getSubmissionSearchIndex()->submissionMetadataChanged($submission);
-            Application::getSubmissionSearchIndex()->submissionFilesChanged($submission);
-            Application::getSubmissionSearchDAO()->flushCache();
-            Application::getSubmissionSearchIndex()->submissionChangesFinished();
+        if (itsPublished) {
+            dispatch(new UpdateSubmissionSearchJob($submission->getId()));
```

Comment:
I think this should tie in more closely with @Vitaliy-1's work (https://github.com/pkp/pkp-lib/issues/6970 and the `PublishEvent` above) -- i.e. the job should be processed as part of the `PublishEvent`.

Reply:
(There are also other calls to re-index full text that should get the same treatment -- e.g. in `lib/pkp/controllers/api/file/PKPManageFileApiHandler.inc.php`.)